src/Common/src/System/Net/SafeCloseSocket.Windows.cs

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+
+namespace System.Net.Sockets
+{
+#if DEBUG
+    internal sealed partial class SafeCloseSocket : DebugSafeHandleMinusOneIsInvalid
+#else
+    internal sealed partial class SafeCloseSocket : SafeHandleMinusOneIsInvalid
+#endif
+    {
+        private ThreadPoolBoundHandle _iocpBoundHandle;
+        private object _iocpBindingLock = new object();
+
+        public ThreadPoolBoundHandle IOCPBoundHandle
+        {
+            get
+            {
+                return _iocpBoundHandle;
+            }
+        }
+
+        //
+        // Binds the Socket Win32 Handle to the ThreadPool's CompletionPort.
+        //
+        public ThreadPoolBoundHandle GetOrAllocateThreadPoolBoundHandle()
+        {
+            if (_released)
+            {
+                // Keep the exception message pointing at the external type.
+                throw new ObjectDisposedException(typeof(Socket).FullName);
+            }
+
+            //
+            // Check to see if the socket native m_Handle is already
+            // bound to the ThreadPool's completion port.
+            //
+            if (_iocpBoundHandle == null)
+            {
+                lock (_iocpBindingLock)
+                {
+                    if (_iocpBoundHandle == null)
+                    {
+                        //
+                        // Bind the socket native m_Handle to the ThreadPool.
+                        //
+                        GlobalLog.Print("SafeCloseSocket#" + Logging.HashString(this) + "::BindToCompletionPort() calling ThreadPool.BindHandle()");
+
+                        try
+                        {
+                            // The handle (this) may have been already released:
+                            // E.g.: The socket has been disposed in the main thread. A completion callback may
+                            //       attempt starting another operation.
+                            _iocpBoundHandle = ThreadPoolBoundHandle.BindHandle(this);
+                        }
+                        catch (Exception exception)
+                        {
+                            if (ExceptionCheck.IsFatal(exception)) throw;
+                            CloseAsIs();
+                            throw;
```

Comment:
(I just realized this code is just being moved... I'll skip over these Windows files)